File: Semana_3/qa/typescript/src/profile-cache.ts
const apiBaseUrl = (process.env.API_BASE_URL ?? 'http://localhost:8000').replace(/\/$/, '');

async function profile(mode: 'cached' | 'nocache'): Promise<number[]> {
  const samples: number[] = [];
  for (let i = 0; i < 5; i += 1) {
    const started = performance.now();
    const response = await fetch(`${apiBaseUrl}/api/lab?mode=${mode}&locale=es`);
    if (!response.ok) {
      throw new Error(`TYPESCRIPT profile_cache failed for mode=${mode}`);
    }
    samples.push(performance.now() - started);
  }
  return samples;
}

function average(values: number[]): number {
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

const cached = await profile('cached');
const nocache = await profile('nocache');

const cachedAvg = average(cached);
const nocacheAvg = average(nocache);

if (cachedAvg >= nocacheAvg) {
  throw new Error('TYPESCRIPT profile_cache expected cached avg lower than nocache avg');
}

console.log('TYPESCRIPT profile_cache OK');
console.log(`cached avg=${cachedAvg.toFixed(2)} ms`);
console.log(`nocache avg=${nocacheAvg.toFixed(2)} ms`);

File: Semana_3/qa/java/src/main/java/com/pulselab/qa/ProfileCache.java
package com.pulselab.qa;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class ProfileCache {
    public static void main(String[] args) throws Exception {
        String baseUrl = System.getenv().getOrDefault("API_BASE_URL", "http://localhost:8000");
        HttpClient client = HttpClient.newHttpClient();

        double cachedAvg = averageMode(client, baseUrl, "cached");
        double nocacheAvg = averageMode(client, baseUrl, "nocache");

        if (cachedAvg >= nocacheAvg) {
            throw new IllegalStateException("JAVA profile_cache expected cached avg lower than nocache avg");
        }

        System.out.println("JAVA profile_cache OK");
        System.out.printf("cached avg=%.2f ms%n", cachedAvg);
        System.out.printf("nocache avg=%.2f ms%n", nocacheAvg);
    }

    private static double averageMode(HttpClient client, String baseUrl, String mode) throws IOException, InterruptedException {
        double sum = 0;
        for (int i = 0; i < 5; i++) {
            long started = System.nanoTime();
            get(client, baseUrl + "/api/lab?mode=" + mode + "&locale=es");
            sum += (System.nanoTime() - started) / 1_000_000.0;
        }
        return sum / 5;
    }

    private static void get(HttpClient client, String url) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(url)).GET().build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new IllegalStateException("Request failed: " + url + " status=" + response.statusCode());
        }
    }
}
